### ISSUE TITLE
Run tests/setup.exp to configure server (and fix segfaults)

### ISSUE THREAD

**Issue by ddugovic** [ddugovic/capablanca#2]
After doing `make` and `make install` to `/usr/local/chessd`, upon running the server and connecting any client the following occurs (after removing from `ficsmain.c` the broken segfault signal handler):

```
chessd@dugovic-host:/usr/local/chessd$ gdb bin/chessd
...

(gdb) run -f -p 5000
Starting program: /usr/local/chessd/bin/chessd -f -p 5000
CHESSD: Initialized on port 5000
Sat Nov 12, 16:04 CST 2016 CHESSD: Loaded 166 commands (admin=46 normal=120)
Sat Nov 12, 16:04 CST 2016 Could not open file "/usr/local/chessd/data/stats/newratingsV2_data" [No such file or directory]
Sat Nov 12, 16:04 CST 2016 CHESSD: Can't read ratings data!
Sat Nov 12, 16:04 CST 2016 CHESSD: Loaded books
Sat Nov 12, 16:04 CST 2016 Too many players connected!

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff75a9fcf in memset (__len=784, __ch=0, __dest=0xfffffffffffffcf0) at /usr/include/x86_64-linux-gnu/bits/string3.h:90
90        return __builtin___memset_chk (__dest, __ch, __len, __bos0 (__dest));
(gdb) bt
#0  0x00007ffff75a9fcf in memset (__len=784, __ch=0, __dest=0xfffffffffffffcf0) at /usr/include/x86_64-linux-gnu/bits/string3.h:90
#1  player_zero (p=p@entry=-1) at playerdb.c:81
#2  0x00007ffff75aa523 in player_new () at playerdb.c:60
#3  0x00007ffff7589e85 in process_new_connection (fd=fd@entry=6, fromHost=...) at command.c:823
#4  0x00007ffff75a9402 in select_loop () at network.c:574
#5  0x0000000000401102 in main_event_loop () at ficsmain.c:90
#6  main (argc=4, argv=0x7fffffffe4f8) at ficsmain.c:218
(gdb) q
A debugging session is active.

        Inferior 1 [process 22905] will be killed.

Quit anyway? (y or n) y
```

**Comment by ddugovic** [comment]
I don't understand this codebase very well but somehow I prevented the segmentation fault.

**Comment by ddugovic** [comment]
Removing `malloc.h` and `malloc.c` and using standard library functions (thus breaking "hot reload" which frees allocated memory) yields:

```
chessd@dugovic-host:/usr/local/chessd$ bin/chessd -f -p 5000
CHESSD: Initialized on port 5000
tdb_open_ex: could not open file /usr/local/chessd/config.tdb: No such file or directory
Sat Nov 12, 19:35 CST 2016 CHESSD: Loaded 166 commands (admin=46 normal=120)
Sat Nov 12, 19:35 CST 2016 CHESSD: Loaded books
IDLE_TIMEOUT=36000
LOGIN_TIMEOUT=300
MAX_PLAYERS=5000
DEFAULT_RATING=0
DEFAULT_RD=350
DEFAULT_TIME=2
DEFAULT_INCREMENT=12
Sat Nov 12, 19:35 CST 2016 Could not open file "/usr/local/chessd/data/lists/ban" [No such file or directory]
Sat Nov 12, 19:35 CST 2016 Could not open file "/usr/local/chessd/data/lists/admin" [No such file or directory]
MAX_PLAYER=5000
Sat Nov 12, 19:35 CST 2016 Could not open file "/usr/local/chessd/players/a/admin" [No such file or directory]
GUEST_PREFIX_ONLY=0
Sat Nov 12, 19:35 CST 2016 Could not open file "/usr/local/chessd/data/lists/filter" [No such file or directory]
Sat Nov 12, 19:35 CST 2016 Could not open file "/usr/local/chessd/data/adverts/0" [No such file or directory]
Sat Nov 12, 19:35 CST 2016 Could not open file "/usr/local/chessd/data/lists/removedcom" [No such file or directory]
Sat Nov 12, 19:35 CST 2016 Could not open file "/usr/local/chessd/players/a/admin" [No such file or directory]
bin/chessd: symbol lookup error: /usr/local/chessd/lib/chessd.so: undefined symbol: crypt
```

**Comment by ddugovic** [comment]
```
chessd@dugovic-host:/usr/local/chessd$ bin/chessd -f -p 5000
CHESSD: Initialized on port 5000
Sat Nov 12, 21:05 CST 2016 CHESSD: Loaded 166 commands (admin=46 normal=120)
Sat Nov 12, 21:05 CST 2016 CHESSD: Loaded books
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/lists/ban" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/lists/admin" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/adverts/0" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/stats/player_data/a/admin.messages" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/lists/computer" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/lists/fm" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/lists/im" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/lists/gm" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/lists/wgm" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/lists/teams" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/lists/blind" [No such file or directory]
Sat Nov 12, 21:06 CST 2016 Could not open file "/usr/local/chessd/data/lists/removedcom" [No such file or directory]
```

**Comment by ddugovic** [comment]
Server seems to be working now.

**Comment by ddugovic** [comment]
Maybe it is possible to safely introduce a SIGSEGV handler.

**Comment by ddugovic** [comment]
```
chessd@dugovic-host:~$ /usr/local/chessd/bin/chessd -f -T /home/chessd/capablanca/timeseal/timeseal_decoder-Linux-ELF-2.4 -p 5000
CHESSD: Initialized on port 5000
Tue Jan 10, 07:38 CST 2017 CHESSD: Loaded 168 commands (admin=46 normal=122)
Tue Jan 10, 07:38 CST 2017 CHESSD: Loaded books
Tue Jan 10, 07:38 CST 2017 malloc corruption ptr=0x18ff120 v1=0 size=33
Tue Jan 10, 07:38 CST 2017 bad free at config.c(137)!
Tue Jan 10, 07:38 CST 2017 malloc corruption ptr=0x18fdff0 v1=0 size=33
Tue Jan 10, 07:38 CST 2017 bad free at config.c(137)!
Tue Jan 10, 07:38 CST 2017 malloc corruption ptr=0x18fe010 v1=0 size=33
Tue Jan 10, 07:38 CST 2017 bad free at config.c(137)!
```
etc.

**Comment by HGMuller** [comment]
I only started reading up on your list of commits yesterday, because of the recent trouble I experienced with the version I operate. (I pretty much operate this ICS based on the "if it aint broken, don't fix it" principle.) So forgive me if I am still confused about what you are trying to achieve with the various commits.
I noticed you removed the -m32 flag from the compile options. It was I who added that flag, when I switched from a 32-bit laptop to a rented 64-bit VPS as host machine, because the native 64-bit compile swamped me with run-time errors. It is not at all clear whether the code is 64-bit compatible. When it stores pointers in integers it certainly would not be, and the code is in general so dirty that I would not exclude it does that kind of thing. Without casts this should give warnings, but compiling the ICS produces tons of warnings, which simply swamps any useful ones. (And even then, it could use a cast, and give no warning at all.) So I would be very hesitant to make a change so far reaching as changing the data formats from 32 to 64 bit standards, without first addressing all existing warnings. Many of the problems you experience here with pointer corruption could be caused by clipping the pointers to 32 bit. Note that the pointer values in the complaints by 'free' above all fit in 32 bits.

**Comment by ddugovic** [comment]
No worries.  On the new `segfault` branch I reverted (maybe unnecessary) changes which had disabled the segfault handler and "hot reload" malloc magic, since on my current `master` it is difficult to fix bugs.

Sorry that I grossly underestimated how stable the 32-bit binaries are and thanks for the excellent advice!  I will re-add `-m32` and advise @iw6dgm that [64-bit binaries are unsupported](http://github.com/ddugovic/capablanca/issues/1#issuecomment-231526959).

**Comment by HGMuller** [comment]
FYI: I have created a new branch 'warnings' in the hgm.nubati.net repository, where I have silenced all compiler warnings. To achieve this I did suppress the warning about non-literal format strings in print statements by a compiler option, though. (-Wno-format-security). I guess inserting an extra "%s" format argument would be the proper way to solve this. But it occurred rather often, mostly with strings that the ICS built itself (like player handles with titles), so that there isn't really a secutity risk. Perhaps I will still fix this later. Most fixes are purely cosmetic anyway; e.g. the complaints about ignored results (also very frequent) were silenced by putting the result in an if clause, but this executes an empty statement on an error. I guess the ICS could print a message in the error log for this instead.

The main purpose was to get rid of the 'warning noise'.

**Comment by ddugovic** [comment]
If I run the `segfault` branch without `-f` the only problems I see are the exit code 1 and the absence of a log file (as far as I can tell).  Perhaps I am overlooking something:

```
● chessd.service - Job that runs the chessd daemon
   Loaded: loaded (/lib/systemd/system/chessd.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2017-01-16 23:41:59 CST; 6h ago
     Docs: man:chessd(1)
  Process: 5765 ExecStart=/usr/local/chessd/bin/chessd -p 5000 -T /home/chessd/capablanca/timeseal/timeseal_decoder-Linux-ELF-2.4 (code=exited, statu
 Main PID: 21925 (code=exited, status=1/FAILURE)
    Tasks: 2
   Memory: 1.3M
      CPU: 1.680s
   CGroup: /system.slice/chessd.service
           ├─5767 [timeseal]
           └─5768 /usr/local/chessd/bin/chessd -p 5000 -T /home/chessd/capablanca/timeseal/timeseal_decoder-Linux-ELF-2.4
```

**Comment by HGMuller** [comment]
I have no experience running systemd daemons, so I have no idea where the output on stderr of the launched program is to go. But chessd should print something on stderr, whether it succeeds or fails. Perhaps it goes to /dev/null by default, and you have to redirect it by hand if you want it to go somewhere else.

That two PIDs are mentioned at the end suggests it succeeded. Timeseal would never be forked off if the initial chessd process (without -f option) would not have been able to fork off a child to do the real work. I doubt anyway that the exit code of the child could be reported here, But when forking off a child succeeds, the origial process immediately does exit(0), so it remains a mystery how you could ever get status=1.

This is a 64-bit compile? I tried to compile without the -m32 flag last week, to see if I could reproduce the strange pointer corruptions you seemed to have. But the result segfaulted immediately in the library routine waitpid(), apparently calling a routine at address 0. It is hard to see if everything worked when we cannot see what it prints on stderr.

**Comment by ddugovic** [comment]
OK, I will experiment without systemd and expect something in stderr.  (Most confusing was that I changed the `exit` codes in `ficsmain.c` and somehow I still see status=1.)

This 32-bit binary was compiled and installed using:

```
chessd@dugovic-host:~/capablanca/src$ uname -a
Linux dugovic-host 4.4.0-59-generic #80-Ubuntu SMP Fri Jan 6 17:47:47 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
chessd@dugovic-host:~/capablanca/src$ ./configure --host=i686-linux-gnu "CFLAGS=-g -ggdb -m32" "CXXFLAGS=-g -ggdb -m32" "LDFLAGS=-m32" "LTCC=gcc -g -ggdb -m32" && make clean && make && make install
```

**Comment by HGMuller** [comment]
Ah OK, you forced the 32-bit by hand. I thought it was 64-bit, because in the Makefile.in of the segfault branch the -m32 is no longer present.

For 32-bit the ICS should work fine. Last Saturday I ran a tournament with 44 engines participating. I don't know if you made any tricky changes in this branch that could jeopardize the stability. The reason it suddenly became crash-prone in October must have been that one of the regular visitors set his 'avail' variable to 1; the bug I fixed that seems to have cured everything was in code that would only be executed when someone had to be informed of players becoming available again when their game ended. I have not been able to identify which user this could be. (First I thought it would be mamer, but that also seemed to have avail=0).
